### PR TITLE
Author

### PR DESCRIPTION
**Pull request by mraypold**
This is a significant refactor of the author friend/follower system and makes it easier to integrate foreign friend/follower relations.

Changes:
- Added a CachedAuthor model. This model holds both home node Authors and foreign node Authors.
- Removed preceding author in friend/follow endpoints as per specs.
- Friendrequest query matches specs exactly.
- One of two friends query from example specs implemented.
- Prior friend/follower code has been moved to generics/mixins.
- CachedAuthors are created for each Author upon registration. (Necessary for friend/follower relations)
- Internal endpoints for followers will return displayname, host, id and url, instead of the guid that is returned for external endpoints.

Known problems:
- The old permission system will not work with this refactor and needs to be switched over. (Should be as simple as querying the many-to-many relations in the author table instead of the FriendRelations and FollowerRelations tables)
- Old models have yet to be deleted as it will break a lot of existing code. Will come in the next PR.
- Some controller logic is mixed with model logic in Author models.
- Friend/follower queries to foreign nodes not implemented. Marked TODO.
- Tests in test_follower_api can be simplified with some scaffolding.

### CODE REVIEW

**Comment by lanre-ade** [comment]
:+1:

File: author_api/models.py [hunk 1]
```diff
@@ -30,9 +58,80 @@ def host(self):
...
 
+    objects = AuthorManager()
+
+    # mixing controller logic with model logic unfortunately
+    def add_follower(self, follower):
+        """Create a follower from an Author/CachedAuthor model"""
+        _cached = CachedAuthor.objects.get(id = follower.id)
+
+        # Prevent duplicate entries
+        if not self.followers.filter(id = follower.id):
```
Comment (derekdowling) [inline author_api/models.py]: Nitpicky, but I suspect doing a get is faster here. Is there a specific reason to use filters in this instance? Lack of exception throwing?

File: author_api/models.py [hunk 2]
```diff
@@ -30,9 +58,80 @@ def host(self):
...
+            #   1. The follower is actually foreign hosted
+            # TODO
+            # To determine if we can add friendship, must query foreign node
+            # self.add_friend(newfriend)
+            pass
+
+    # Call add_follower instead
+    def add_friend(self, follower):
+        """Create a friend from an author model"""
+        _cached = CachedAuthor.objects.get(id = follower.id)
```
Comment (derekdowling) [inline author_api/models.py]: I think this needs a try/except.

File: author_api/models.py [hunk 3]
```diff
@@ -30,9 +58,80 @@ def host(self):
     def url(self):
       return self.host + 'author/' + str(self.id)
 
+    objects = AuthorManager()
```
Comment (derekdowling) [inline author_api/models.py:63]: I'm a bit confused as to what this declaration does.

Comment (mraypold) [inline author_api/models.py:63]: This allows the custom queries to be used from AuthorQuerySet and AuthorManager. Basically it extends the Author.objects query manager to allow an areFriends query defined above.

File: rest_api/urls.py [hunk 4]
```diff
@@ -14,8 +14,9 @@
 urlpatterns = patterns('',
     url(r'^grappelli/', include('grappelli.urls')),
     url(r'^admin/', include(admin.site.urls)),
-    url(r'^author/', include('author_api.urls')),
+    # url(r'^author/', include('author_api.urls')),
```
Comment (derekdowling) [inline rest_api/urls.py]: Delete this if we aren't using it anymore.

File: author_api/models.py (hunk 3 above)
Comment (derekdowling) [inline author_api/models.py:63]: Oooh, so basically, you're overriding the default Author.objects.blah() syntax so you can use the helper? Cool! Can you add that as a comment then? It's not immediately obvious.

File: author_api/urls.py [hunk 5]
```diff
@@ -4,28 +4,32 @@
...
 
     # POST /logout
-    url(r'^logout/?$', viewsauth.Logout.as_view(), name='logout'),
+    url(r'^author/logout/?$', viewsauth.Logout.as_view(), name='logout'),
 
     # GET POST /author/profile
-    url(r'^profile/?$', viewsauth.AuthorProfile.as_view(), name='profile'),
+    url(r'^author/profile/?$', viewsauth.AuthorProfile.as_view(), name='profile'),
 
     # GET /author/friends/:username
```
Comment (derekdowling) [inline author_api/urls.py:15]: This is missing /author? Is the new spec correct? If so this comment should be updated.

File: author_api/views.py [hunk 6]
```diff
@@ -73,130 +56,152 @@ def get(self, request, *args, **kwargs):
...
+        Remove the follower and the associated friend for the given author
+        Takes:
+            author: Author model
+            follower: CachedAuthor model
+        """
+        # This will remove the friendship. Follower/Friendship are dependents
+        author.remove_follower(follower)
+
+# Can potentially be moved to a viewset and router
+class CRDFollowers(
```
Comment (derekdowling) [inline author_api/views.py]: You should consider renaming this to FollowersView or something of the like. The CRD part should be evident in the mixins and using that in the name distracts from what it is and doesn't follow any convention we use anywhere else.

**Comment by derekdowling** [comment]
Okay, other than a few pieces of feedback this looks good. Addess those and then GTG!

File: author_api/urls.py (hunk 5 above)
Comment (mraypold) [inline author_api/urls.py:15]: I'm not following. This is the same url as previously. However, author used to be prefixed to all author_api urls when called from rest_api/urls.py. Now rest_api/urls.py doesn't prefix author and its prefixed within author_api as some endpoints do not need /author.

File: author_api/models.py [hunk 7]
```diff
@@ -30,9 +58,80 @@ def host(self):
...
+        except:
+            # This will fail if:
+            #   1. The follower is actually foreign hosted
+            # TODO
+            # To determine if we can add friendship, must query foreign node
+            # self.add_friend(newfriend)
+            pass
+
+    # Call add_follower instead
+    def add_friend(self, follower):
```
Comment (derekdowling) [inline author_api/models.py:91]: Also, for friend related functions, aka this one and removeFriend, the "follower" argument should be "friend".

**Comment by xkb** [comment]
I'm not sure if I fully understand the purpose of `CachedAuthor` and its _name_. It sounds like the intention is to have proxy object that returns persisted authors or foreign authors? That should probably be done in-memory rather than having an entire db table by itself.

I know that this is also ground works for upcoming changes to communicate with foreign nodes but keep in mind that we're expected to have inconsistent state between nodes. Cached state, if you're planning on keeping it is over kill.

Either way, functionally this PR pushes us in the right direction :+1:

**Comment by derekdowling** [comment]
@KarimBaaba I think it's just poorly named. Basically it just provides a bunch of helpers surrounding the id and host of friends/followersso that we have to do less construction of urls in other places. This is basically just storing the most basic amount of information we need to look up friend/follower details.

**Pull request merged**